simba/costs.py

```diff
@@ -100,21 +100,41 @@ def calculate_costs(c_params, scenario, schedule, args):
         costs["c_stat_storage"] * c_params["stationary_storage"]["c_maint_stat_storage_per_year"])
 
     # CHARGING INFRASTRUCTURE
-    cs = schedule.scenario["components"]["charging_stations"]
-    # depot charging stations - each charging bus generates one CS
-    for csID in cs.values():
-        if csID["type"] == "deps":
-            costs["c_cs"] += c_params["cs"]["capex_deps_per_kW"] * csID["max_power"]
-    # opportunity charging stations - nr of CS depend on max nr of simultaneously occupied CS
-    for gcID in gcs:
-        if schedule.stations[gcID]["type"] == "opps":
-            gc_timeseries = getattr(scenario, f"{gcID}_timeseries")
-            costs["c_cs"] += (
-                c_params["cs"]["capex_opps_per_kW"] *
-                # get cs power of this opps
-                schedule.stations[gcID].get("cs_power_opps", vars(args)["cs_power_opps"]) *
-                # get max. nr of occupied CS per grid connector
-                max(gc_timeseries["# CS in use [-]"]))
+    stations = schedule.scenario["components"]["charging_stations"]
+    for gcID, station in schedule.stations.items():
+        # check if electrified station from list is actually used in simulation
+        if gcID in gc_in_use:
+            # depot charging station - each charging bus generates one CS
+            if station["type"] == "deps":
+                if station["n_charging_stations"] is not None:
+                    # get nr of CS in use
+                    n_cs = min(sum(gcID == cs["parent"] for cs in stations.values()),
+                               station["n_charging_stations"])
+                    # get max. defined power for deps or else max. power at deps in general
+                    max_defined_power = max(
+                        station.get("cs_power_deps_depb", vars(args)["cs_power_deps_depb"]),
+                        station.get("cs_power_deps_oppb", vars(args)["cs_power_deps_oppb"])
+                    )
+                    # calculate costs with nr of CS at electrified station and its defined max power
+                    costs["c_cs"] += c_params["cs"]["capex_deps_per_kW"] * n_cs * max_defined_power
+                else:
+                    # get sum of installed power at electrified stations without predefined nr of CS
+                    nr_cs_at_station = [cs for cs in stations.values() if cs["parent"] == gcID]
```

Comment:
This seems to be independent of whether `n_charging_stations` is set or not. So, move this line before line 109, the you can use it to compute n_cs in line 111 as well: `n_cs = min(len(cs_at_station), station["n_charging_stations"])` -- of course, the `min` only makes sense if there can be more CS at a station than n_charging_stations. (Also renamed nr_cs_at_station in this example, because it is a list of actual stations and not just the count).

Reply:
Done. The `min` needs to stay, since `n_charging_stations` is respected by the simulation but in the list `cs_at_station` still for every bus there is a charging station generated.